src/str.c

```diff
@@ -677,6 +687,212 @@ DEFINE_PRIMITIVE("string-append", string_append, vsubr, (int argc, SCM* argv))
   return z;
 }
 
+/*
+<doc SRFI-118 string-append!
+ * (string-append! string ...)
+ *
+ * Extends string by appending each value (in order) to the end of string.
+ * A value can be a character or a string.
+ *
+ * It is guaranteed that string-append! will return the same object that
+ * was passed to it as first argument, whose size may be larger.
+doc>
+ */
+DEFINE_PRIMITIVE("string-append!", string_append_adjsize, vsubr, (int argc, SCM* argv))
+{
+  unsigned long i, total_size, total_length;
+  char *q; /* first string */
+  char *p; /* others (chars or strings) */
+
+  if (argc==0) STk_error("incorrect number of arguments (%d)", argc);
+  
+  /* First argument MUST be a string: */
+  if (STRINGP(argv[0]))
+      total_size = STRING_SIZE(argv[0]);
+  else
+      error_bad_string(argv[0]);
+  
+  /* Compute total length of resulting string */
+  for (i = 1; i < argc; i++) {
+    p = argv[-i];
+    if (STRINGP(p))
+        total_size += STRING_SIZE(p);
+    else if (CHARACTERP(p))
+        total_size+=STk_utf8_char_bytes_needed(CHARACTER_VAL(p));
+    else
+     error_bad_string_or_character(p);
+  }
+
+  q = argv[0];
+
+  /* Reallocate space in first argument string: */
+  STRING_CHARS(q) = STk_must_realloc(STRING_CHARS(q), total_size);
+
+  /* point to the last character of first string: */
+  p = (STRING_CHARS(q)+STRING_SIZE(q));
+
+  STRING_SPACE(q) = STRING_SIZE(q) = total_size;
+      
+  /* copy strings. i starts at ONE, not zero, since the first
+     string need not be copied. */
+  for (total_length=STRING_LENGTH(q), i=1; i < argc; i++) {
+    if (STRINGP(*(argv-i))) {
+      memcpy(p, STRING_CHARS(argv[-i]), (unsigned int) STRING_SIZE(argv[-i]));
+      p += STRING_SIZE(*(argv-i));
+      total_length += STRING_LENGTH(*(argv-i));
+    } else { /* we have already checked, it is a char! */
+      p += STk_char2utf8(CHARACTER_VAL(*(argv-i)),p);
+      /*      p += STk_utf8_char_bytes_needed(CHARACTER_VAL(p));*/
+      total_length ++;
+    }
+  }
+
+  STRING_LENGTH(q) = total_length;
+  fflush(stdout);
```

Comment:
could be deleted?

Reply:
Ha ha! Sorry, that's from debugging, I forgot to delete. Will do that.